### ISSUE TITLE
link time error building Salmon 1.0.0

### ISSUE THREAD

**Issue by patoddh** [COMBINE-lab/salmon#455]
This is an installation problem.

I am getting a link-time error building Salmon 1.0.0 with GCC 7.3.0 and CMake 3.12.1. The error occurs immediately after "[100%] Linking CXX executable salmon". The OS is RHEL6.

Here's the output:

```
/software/Core/GCCcore/7.3.0/include/c++/7.3.0/bits/shared_ptr_base.h:522: error: undefined reference to 'vtable for std::_Sp_counted_ptr_inplace<spdlog::logger, std::allocator<spdlog::logger>, (__gnu_cxx::_Lock_policy)2>'
/software/Compiler/GCCcore/7.3.0/binutils/2.30/bin/ld.gold: the vtable symbol may be undefined because the class is missing its key function
/software/Compiler/GCCcore/7.3.0/binutils/2.30/bin/ld.gold: the symbol should have been defined by a plugin
/software/Core/GCCcore/7.3.0/include/c++/7.3.0/bits/shared_ptr_base.h:522: error: undefined reference to 'vtable for std::_Sp_counted_ptr_inplace<spdlog::async_logger, std::allocator<spdlog::async_logger>, (__gnu_cxx::_Lock_policy)2>'
/software/Compiler/GCCcore/7.3.0/binutils/2.30/bin/ld.gold: the vtable symbol may be undefined because the class is missing its key function
/software/Compiler/GCCcore/7.3.0/binutils/2.30/bin/ld.gold: the symbol should have been defined by a plugin
```

Googling this error implies an issue with the code, not a missing library.

**Comment by avapirev** [comment]
Error confirmed. OS RHEL7, GCC 6.4.0, CMake 3.13.x

**Comment by cljacobs** [comment]
I've run into this as well, but's proving tricky to track down. I'm by no mean a expert on build chains (or a C/C++ for that matter), but as far as I can figure, this issue seems specific to RedHat systems.

I get exactly the same linking error on RH6 and RH7 using any GCC compiler I have access to on those systems (4.8.5, 5.2.0, 7.2.0). Compiling on an Arch system with GCC 9.2.0 (glibc 2.30) sees no issue. Unfortunately I don't have easy access to the same compiler versions on both systems. I'm compiling GCC 5.2.0 (with glibc2.28) on the Arch system to test this now, but it's going to be a little while before I even know if I have a working toolchain that can use it.

**Comment by rob-p** [comment]
Thanks for the details @cljacobs.  We'll see if we can get a Docker image up to reproduce this under RH7.  Our development machine is ubuntu based, and our CI is CentOS.  It also builds on the environment used by bioconda.  So it looks like we'll need a RH image to reproduce this.  Out of curiosity, does anything happen differently if you pass `-DNO_IPO=TRUE` during the `cmake` configure step?  That disables interprocedural optimization (whole program link-time optimization).

**Comment by cljacobs** [comment]
Oh yes. `-DNO_IPO=TRUE` has fixed it for me. RH7/GCC7.2.0, at least.

**Comment by patoddh** [comment]
That fixed it for me on RHEL6 and GCC 7.3.0. Thanks!

**Comment by avapirev** [comment]
That tip fixed it for me too: OS RHEL7, GCC 6.4.0, CMake 3.13.x
@cljacobs Thanks a lot.

**Comment by k3yavi** [comment]
The issue seems to be fixed but feel free to reopen if the problem still remains.
